File: ESC190/lab07/circular_queue.py
class CircularQueue:
    def __init__(self, sz, start):
        self.sz = sz
        self.queue = [None] * sz
        self.start = start
        self.end = start
        self.cur_size = 0
        
    def enqueue(self, item):
        if self.queue[self.start] is None:
            self.queue[self.start] = item
            self.cur_size += 1
            self.end = self.start
        elif self.cur_size != self.sz:    
            ind = -1
            if self.end + 1 != self.sz:
                self.queue[self.end+1] = item
                ind = self.end + 1
            else:
                for i in range(0, self.start):
                    if self.queue[i] is None:
                        self.queue[i] = item
                        ind = i
                        break
            self.end = ind
            self.cur_size += 1
        else:
            print("Queue is full")
    
    def dequeue(self):
        if self.cur_size != 0:
            item = self.queue[self.start]
            self.queue[self.start] = None
            if self.start + 1 != self.sz:
                self.start += 1
            else:
                self.start = 0
            self.cur_size -= 1
            return item
        else:
            print("Queue is empty")
    
    def __repr__(self):
        out = "Current Queue: ["
        for i in range(0, self.sz-1):
            out += str(self.queue[i]) + ", "
        out += str(self.queue[self.sz-1]) + "]"
        # out += "\tCircularQueue: ["
        # for i in range(self.start, self.sz):
        #     if i == self.end:
        #         out += str(self.queue[i]) + "]"
        #     elif self.queue[i] is not None:
        #         out += str(self.queue[i]) + ", "
        
        # for i in range(0, self.end):
        #     if i == self.end:
        #         out += str(self.queue[i]) + "]"
        #     elif self.queue[i] is not None:
        #         out += str(self.queue[i]) + ", "
        
        return out
    
    def __lt__(self, other):
        for i in range(0, min(self.cur_size, other.cur_size)):
            if self.queue[(self.start + i) % self.sz] < other.queue[(other.start + i) % other.sz]:
                return True
            elif self.queue[(self.start + i) % self.sz] > other.queue[(other.start + i) % other.sz]:
                return False
        return False
        

if __name__ == "__main__":
    q = CircularQueue(5, 1)
    q.enqueue(1)
    q.enqueue(2)
    q.enqueue(3)
    print(q)
    q.dequeue()
    print(q)
    q.enqueue(4)
    print(q)
    q.enqueue(5)
    print(q)
    q.enqueue(6)
    print(q)
    
    m = CircularQueue(5, 2)
    m.enqueue(1)
    m.enqueue(2)
    m.enqueue(3)
    print(q, m)
    print(q < m)

File: ESC190/lab07/circular_queue.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct CircularQueue{
    int *data;
    int front;
    int rear;
    int size;
    int cur_size;
} CircularQueue;

void create_CircularQueue(CircularQueue **p_q, int size, int start)
{
    (*p_q) = (CircularQueue*)malloc(sizeof(CircularQueue));
    (*p_q)->data = (int*)malloc(sizeof(int)*size);
    (*p_q)->front = start;
    (*p_q)->rear = start;
    (*p_q)->size = size;
    (*p_q)->cur_size = 0;

    for (int i = 0; i < size; i++) {
        (*p_q)->data[i] = 0;
    }
}

void enqueue(CircularQueue *q, int new_elem)
{
    if (q->cur_size == q->size) {
        printf("Queue is full!\n");
        q->data = realloc(q->data, sizeof(int)*(q->size + 1));
        q->size++;
        if (q->rear < q->front) {
            for (int i = q->size - 1; i > q->rear; i--) {
                q->data[i] = q->data[i - 1];
            }
            q->front = q->front + 1;
        }
        q->rear = q->rear + 1;
        q->data[q->rear] = new_elem;
        q->cur_size++;
        return;
    }

    if (q->cur_size == 0) {
        q->data[q->rear] = new_elem;
    } else {
        q->rear = (q->rear + 1) % q->size;
        q->data[q->rear] = new_elem;
    }

    q->cur_size++;
}

int dequeue(CircularQueue *q)
{
    if (q->cur_size == 0) {
        printf("Queue is empty!\n");
        return -1;
    }

    int val = q->data[q->front];
    q->data[q->front] = 0;
    q->front = (q->front + 1) % q->size;
    q->cur_size--;

    return val;
}

void print_CircularQueue(CircularQueue *q)
{
    printf("Circular Queue: ");
    for (int j = 0; j < q->size; j++) {
        printf("%d ", q->data[j]);
    }
    printf("\n");
}

void main()
{
    CircularQueue *q;
    create_CircularQueue(&q, 5, 1);
    enqueue(q, 1);
    enqueue(q, 2);
    enqueue(q, 3);
    enqueue(q, 4);
    print_CircularQueue(q);
    enqueue(q, 5);
    printf("Dequeue: %d\n", dequeue(q));
    print_CircularQueue(q);
    enqueue(q, 6);
    print_CircularQueue(q);
    enqueue(q, 7);
    print_CircularQueue(q);
    free(q->data);
    free(q);
}